### PR TITLE
Adding safe casts header and implementation

### PR DESCRIPTION
**Pull request by stephenramm**
@lcweaver 
Added an implementation very similar to the khronos example implementation in the docs repo
@lcweaver I have added conformance tests, it is ready for review.

### CODE REVIEW

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+TEST(tivxSafeCasts, testCastGetAsSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+
+    for (i = 0; i < UPCAST_NUM_TYPES; ++i)
+    {
+        vx_enum type = upcast_type_ids[i];
+        const char *type_name = enumToString(type);
+        printf("\nTesting vxCastRefAs and vxGetRefAs for %s\n", type_name);
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply conformance_tests/test_tiovx/test_safe_casts.c:641]: Not strictly required - was for information only. I will remove.

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+        }
+    }
+    vxReleaseGraph(&graph);
+}
+
+/* Check basic operation of castAs and GetAs */
+TEST(tivxSafeCasts, testCastGetAsSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply conformance_tests/test_tiovx/test_safe_casts.c:635]: will do

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+        {
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, getAsSomeType(graph, type, SUCCESS));
+            EXPECT_NE_VX_STATUS(VX_SUCCESS, getAsSomeType(graph, type, TYPEERROR));
+            EXPECT_NE_VX_STATUS(VX_SUCCESS, getAsSomeType(graph, type, NULLREF));
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, castAsSomeType(graph, type, SUCCESS));
+            EXPECT_NE_VX_STATUS(VX_SUCCESS, castAsSomeType(graph, type, TYPEERROR));
+            EXPECT_NE_VX_STATUS(VX_SUCCESS, castAsSomeType(graph, type, NULLREF));
+        }
+    }
+    vxReleaseGraph(&graph);
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply conformance_tests/test_tiovx/test_safe_casts.c:652]: will do

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+TEST(tivxSafeCasts, testCastGetFromSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+
+    for (i = 0; i < UPCAST_NUM_TYPES; ++i)
+    {
+        vx_enum type = upcast_type_ids[i];
+        const char *type_name = enumToString(type);
+        printf("\nTesting vxCastRefFrom and vxGetRefFrom for %s\n", type_name);
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply conformance_tests/test_tiovx/test_safe_casts.c:599]: Not strictly required - was for information only. I will remove.

File: source/framework/vx_safe_casts.c [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+{
+    vx_status local_status = VX_SUCCESS;
+    vx_reference retref = ref;
+    if ((NULL != ref) &&
+        (ref->magic == TIVX_MAGIC) &&
+            ownIsValidContext(ref->context))
+    {
+        if (ref->type == type)
+        {
+            ownIncrementReference(ref, VX_EXTERNAL);
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply source/framework/vx_safe_casts.c:139]: OK

File: source/framework/vx_safe_casts.c [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+__DEFINE_SAFE_CASTS__(user_data_object, UserDataObject, USER_DATA_OBJECT)
+#endif
+
+vx_reference getRefAs(vx_reference ref, vx_enum type, vx_status *status)
+{
+    vx_status local_status = VX_SUCCESS;
+    vx_reference retref = ref;
+    if ((NULL != ref) &&
+        (ref->magic == TIVX_MAGIC) &&
+            ownIsValidContext(ref->context))
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply source/framework/vx_safe_casts.c:135]: OK

File: source/framework/vx_safe_casts.c [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    return retref;
+}
+
+vx_reference castRefAs(vx_reference ref, vx_enum type, vx_status *status)
+{
+    vx_status local_status = VX_SUCCESS;
+    vx_reference retref = ref;
+    if ((NULL != ref) &&
+        (ref->magic == TIVX_MAGIC) &&
+         ownIsValidContext(ref->context))
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply source/framework/vx_safe_casts.c:167]: OK

File: source/framework/vx_safe_casts.c [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            vxAddLogEntry(ref, local_status, "vxGetRefAs: Attempt to get reference %p as type %d when actual type is %d\n", ref, type, ref->type);
+            retref = ownGetErrorObject(ref->context, local_status);
+        }
+    }
+    else
+    {
+        local_status = VX_ERROR_INVALID_REFERENCE;
+    }
+    if (status &&
+        VX_SUCCESS == *status)
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply source/framework/vx_safe_casts.c:154]: OK

File: source/framework/vx_safe_casts.c [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            vxAddLogEntry(ref, local_status, "vxCastRefAs: Attempt to cast reference %p as type %d when actual type is %d\n", ref, type, ref->type);
+            retref = ownGetErrorObject(ref->context, local_status);
+        }
+    }
+    else
+    {
+        local_status = VX_ERROR_INVALID_REFERENCE;
+    }
+    if (status &&
+        VX_SUCCESS == *status)
```
[the remark this answers is not in the document]
Comment (stephenramm) [reply source/framework/vx_safe_casts.c:182]: OK

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+    }
+    VX_CALL(vxReleaseGraph(&graph));
+}
+
+/* Check basic operation of castAs and GetAs */
+TEST(tivxSafeCasts, testCastGetAsSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+    ASSERT_VX_OBJECT(graph);)
```
Comment (lcweaver) [inline conformance_tests/test_tiovx/test_safe_casts.c:602]: It looks like the ASSERT_VX_OBJECT requires two arguments, the second being the OpenVX type

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+        default:
+            return vx_false_e;
+   }
+}
+
+TEST(tivxSafeCasts, testCastGetFromSomeType)
+{
+    int i;
+    vx_graph graph = vxCreateGraph(context_->vx_context_);
+    ASSERT_VX_OBJECT(graph);
```
Comment (lcweaver) [inline conformance_tests/test_tiovx/test_safe_casts.c:560]: It looks like the ASSERT_VX_OBJECT requires two arguments, the second being the OpenVX type

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+    for (i = 0; i < UPCAST_NUM_TYPES; ++i)
+    {
+        vx_enum type = upcast_type_ids[i];
+        if (is_supported(type))
+        {
+            /* For both operations we are looking for the correct type and a reference count of 1, except for context, where ref count should be 2 */
+            vx_enum check_type = VX_ERROR_INVALID_TYPE;
+            vx_uint32 ref_count = 0;
+            vx_reference ref = castAsReference(graph, type);
+            ASSERT_VX_OBJECT(ref);
```
Comment (lcweaver) [inline conformance_tests/test_tiovx/test_safe_casts.c:570]: It looks like the ASSERT_VX_OBJECT requires two arguments, the second being the OpenVX type

File: conformance_tests/test_tiovx/test_safe_casts.c [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxQueryReference(ref, VX_REFERENCE_COUNT, &ref_count, sizeof(ref_count)));
+            if (VX_TYPE_CONTEXT == type || VX_TYPE_KERNEL == type)
+            {
+                /* we expect both these to have an extra ref. count */
+                --ref_count;
+            }
+            EXPECT_EQ_INT(ref_count, 1U);
+            EXPECT_EQ_VX_STATUS(VX_SUCCESS, vxReleaseReference(&ref));
+            ref = getAsReference(graph, type);
+            ASSERT_VX_OBJECT(ref);
```
Comment (lcweaver) [inline conformance_tests/test_tiovx/test_safe_casts.c:582]: It looks like the ASSERT_VX_OBJECT requires two arguments, the second being the OpenVX type

File: source/framework/vx_safe_casts.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (lcweaver) [reply source/framework/vx_safe_casts.c:135]: It looks like this is still pending.  I'll leave it marked as unresolved for now in that case

File: source/framework/vx_safe_casts.c [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            vxAddLogEntry(ref, local_status, "vxGetRefAs: Attempt to get reference %p as type %d when actual type is %d\n", ref, type, ref->type);
+            retref = ownGetErrorObject(ref->context, local_status);
+        }
+    }
+    else
+    {
+        local_status = (vx_status)VX_ERROR_INVALID_REFERENCE;
+    }
+    if (status &&
+        (VX_SUCCESS == *status))
```
Comment (lcweaver) [inline source/framework/vx_safe_casts.c:154]: It looks like this VX_SUCCESS still needs a cast

File: source/framework/vx_safe_casts.c [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+            vxAddLogEntry(ref, local_status, "vxCastRefAs: Attempt to cast reference %p as type %d when actual type is %d\n", ref, type, ref->type);
+            retref = ownGetErrorObject(ref->context, local_status);
+        }
+    }
+    else
+    {
+        local_status = (vx_status)VX_ERROR_INVALID_REFERENCE;
+    }
+    if (status &&
+        (VX_SUCCESS == *status))
```
Comment (lcweaver) [inline source/framework/vx_safe_casts.c:183]: It looks like this VX_SUCCESS still needs a cast

**Review by lcweaver — CHANGES_REQUESTED** [review]

**Comment by lcweaver** [comment]
Can you please squash the commits into a single commit with the tag [TIOVX-1540] within the first line of the commit message?

Additionally, please be sure to remove the merge commit when doing so

**Review by lcweaver — CHANGES_REQUESTED** [review]

**Pull request closed without merging**